### DISCUSSION TITLE
what difference between IntVar.setRange and solver.addConstraints

### DISCUSSION THREAD

**Discussion by hlee13** [google/or-tools#4063] [Routing (and legacy CP) questions] [answered]
Version: v9.6.
Language: java

**Which solver are you using **
Routing Solver

When modeling with RoutingModel, use IntVar.setRange(0, upperValue) more efficient than
solver.addConstraints(solver.makeLessOrEqual(IntVar, upperValue)).
What is the reason? Is it due to early pruning by IntVar.setRange?

**Comment by sschnug** [comment, marked as answer]
In general, propagators (build by constraints) are tasked to wait for triggers and then remove values from variable-domains within a fixpoint-iteration loop. This happens a lot during search.

If you can remove those values a-priori, you don't need a propagator anymore and don't pay for it's overhead (listener = wait; propagation-work) as the domain was changed at the root-lvl before any decisions/branchings are made which usually trigger fixpoint-search / propagation.

Every non-toy solver, when seeing `solver.addConstraints(solver.makeLessOrEqual(IntVar, upperValue))` will transform / lift / rewrite this into a domain-change and don't add/post a propagator or get rid of it later during presolve.

So in a good solver, there is no difference but a small presolve-operation which is done ONCE before the core solve-loop aka it doesn't matter.

I'm 99.9% sure, ortools' cp-solver does this presolve transform.

**Reply by lperron** [reply]
CP-SAT does it.
The original cp solver in the routing library does not. I believe it reorders constraint so that the effect is minimal, but I wrote that code 15 years ago.

**Reply by hlee13** [reply]
Thanks to everyone, I am getting deeper and deeper into some of the detailed concepts inside ortools
